x/data/server/msg_server.go

```diff
@@ -38,62 +40,67 @@ func (s serverImpl) AnchorData(goCtx context.Context, request *data.MsgAnchorDat
 	return &data.MsgAnchorDataResponse{Timestamp: timestamp}, nil
 }
 
-func (s serverImpl) anchorCidIfNeeded(ctx sdk.Context, cid []byte) error {
-	if s.anchorTable.Has(ctx, cid) {
+func blockTimestamp(ctx sdk.Context) (*gogotypes.Timestamp, error) {
+	timestamp, err := gogotypes.TimestampProto(ctx.BlockTime())
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "invalid block time")
+	}
+
+	return timestamp, err
+}
+
+func (s serverImpl) anchorCidIfNeeded(ctx sdk.Context, timestamp *gogotypes.Timestamp, cid []byte) error {
+	store := ctx.KVStore(s.storeKey)
+	key := AnchorKey(cid)
+	if store.Has(key) {
 		return nil
 	}
 
-	return s.anchorCid(ctx, nil, cid)
+	return s.anchorCid(ctx, timestamp, cid)
 }
 
 func (s serverImpl) anchorCid(ctx sdk.Context, timestamp *gogotypes.Timestamp, cidBytes []byte) error {
-	err := s.anchorTable.Create(ctx, cidBytes, timestamp)
+	bz, err := timestamp.Marshal()
 	if err != nil {
-		return sdkerrors.Wrap(err, "error anchoring data")
+		return err
 	}
 
+	store := ctx.KVStore(s.storeKey)
+	key := AnchorKey(cidBytes)
+	store.Set(key, bz)
+
 	return ctx.EventManager().EmitTypedEvent(&data.EventAnchorData{Cid: cidBytes})
 }
 
+var emptyBz = []byte{0}
```

Comment:
Any particular reason why this is set outside the function body?

Reply:
Because it doesn't need to be allocated for every call